### DISCUSSION TITLE
Next.js Server action secretly returns promise instead of string?

### DISCUSSION THREAD

**Discussion by ishaangandhi** [vercel/next.js#64342] [Help]
### Summary

I have a function that, according to Typescript, returns a string. However, once I upgraded to Next 14.1.4, it started to return a Promise.

Here is the setup:
```
"use server";
export const getURL = (path: string = "") => {
  const url =
    process.env.NODE_ENV == "development"
      ? "http://localhost:3000/"
      : "https://foo.com/";

  url = url.replace(/\/+$/, "");
  path = path.replace(/^\/+/, "");

  return path ? `${url}/${path}` : url;
};

export async function checkoutWithStripe() {
  ...
  const url = getURL();
  console.log(url);
}
```

## Expected vs. current behavior

What I would have thought would be printed is:
```
http://localhost:3000/
```

But I actually get:
```
Promise {
    'http://localhost:3000/',
    [Symbol(async_id_symbol)]: 771646,
    [Symbol(trigger_async_id_symbol)]: 771307,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: {
      headers: [Getter],
      cookies: [Getter],
      mutableCookies: [Getter],
      draftMode: [Getter]
    },
    [Symbol(kResourceStore)]: {
      isStaticGeneration: false,
      urlPathname: '/begin-trial',
      pagePath: '/(non_hero)/begin-trial/page',
      incrementalCache: [IncrementalCache],
      isRevalidate: false,
      isPrerendering: undefined,
      fetchCache: 'default-no-store',
      isOnDemandRevalidate: false,
      isDraftMode: false,
      postpone: undefined,
      fetchMetrics: [Array],
      revalidate: 0,
      tags: [Array],
      nextFetchId: 3
    },
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: { isAction: true },
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined,
    [Symbol(kResourceStore)]: undefined
  }
```

**Comment by icyJoseph** [comment]
Hi,

Server actions ought to be async function... look at the caveats here:

- https://react.dev/reference/react/use-server#caveats

> Because the underlying network calls are always asynchronous, 'use server' can only be used on async functions.

So you have to await the result, and because of that, you might as well mark it as async.

**Comment by ishaangandhi** [comment]
Ah, thanks. This seems like a bug in Typescript, right?

Also, if I move my getURL function to some typescript file that doesn't have "use server", then am I guaranteed it will return a string and not a promise? Or is it the case that transitively everything called from a use server function could become async?

**Reply by icyJoseph** [reply]
This is like, a place where, Typescript, with the information it is presented with, can't really help you. 

A TypeScript plugin, or linting tool perhaps, should either cause a build error, or editor indicator, that your function is a server action (use server at top of the module, or within the body of the function) and because of that, ought to be marked as async. And that way, force you to await its result.

A server action does a round trip to the server to trigger the actual implementation. So when you, from the client, "call the function" you are actually firing a request, with some information that the server uses to know which code to execute, on the server, then the result is collected and sent to you, that's why it is giving you a promise. Note that the function looks like, it executes within the context of your client, but it doesn't necessarily do that, that's the magic of these.

**Reply by ishaangandhi** [reply]
> This is like, a place where, Typescript, with the information it is presented with, can't really help you.

Thanks, makes sense.

You probably answered my second question, but I didn't follow.

To be more explicit, if I do the following, is it guaranteed that `foo` is a string?
```
// lib/bar.ts
export function bar() {
  return "bar";
}
```

```
// server-action.ts
"use server";
import { bar } from "@lib/bar";

export async function checkout {
  const foo = bar();
}
```

**Reply by icyJoseph** [reply]
Yes it should be, it is checkout that's gonna do the server action magic.
